synapse/lib/cache.py

```diff
@@ -98,6 +84,41 @@ def clear(self):
         self.fifo.clear()
         self.cache.clear()
 
+class LruDict(collections.OrderedDict):
+    '''
+    Maintains the last n accessed keys
+    '''
+    def __init__(self, size=10000):
+        collections.OrderedDict(self)
+        self.maxsize = size
+        self.disabled = not self.maxsize
+
+    def __getitem__(self, key):
+        valu = super().__getitem__(key)
+        self.move_to_end(key)
+        return valu
+
+    def __setitem__(self, key, valu):
+        if self.disabled:
+            return
+        super().__setitem__(key, valu)
+        self.move_to_end(key)
+        if len(self) > self.maxsize:
+            try:
+                # This actually calls __getitem__ above.  Ignore exception raised by move_to_end
+                self.popitem(last=False)
+            except KeyError:
+                pass
+
+    def __delitem__(self, key):
+        '''
+        Ignore attempts to delete keys that may have already been flushed
+        '''
+        try:
```

Comment:
In the case of size=0, does `if self.disable is False: return None` type logic delete quicker than the try/except exception case?

Reply:
may be quicker even to set self.disable to None and then you can do if self.disabled is None checks

Reply:
Using a dict like this is much quicker for the deletion case
```
class LruDict(collections.OrderedDict):
    '''
    Maintains the last n accessed keys
    '''
    def __init__(self, size=10000):
        super().__init__()
        self.maxsize = size
        self.isdisabled = True
        if self.maxsize:
            self.isdisabled = False

    def __getitem__(self, key):
        valu = super().__getitem__(key)
        self.move_to_end(key)
        return valu

    def __setitem__(self, key, valu):
        if self.isdisabled is True:
            return
        super().__setitem__(key, valu)
        self.move_to_end(key)
        if len(self) > self.maxsize:
            try:
                # This actually calls __getitem__ above.  Ignore exception raised by move_to_end
                self.popitem(last=False)
            except KeyError:
                pass

    def __delitem__(self, key):
        '''
        Ignore attempts to delete keys that may have already been flushed
        '''
        if self.isdisabled is True:
            return
        try:
            super().__delitem__(key)
        except KeyError:
            pass
```

Also `.get()` is an interface which does not bump the key value to the end.  Implementing `get()` would keep the class consistent in its behavior.

Reply:
Done